lib/tbot/service_application_tunnel.go

```diff
@@ -0,0 +1,249 @@
+	if l == nil {
+		s.log.DebugContext(ctx, "Opening listener for application tunnel", "listen", s.cfg.Listen)
+		var err error
+		l, err = createListener(ctx, s.log, s.cfg.Listen)
+		if err != nil {
+			return trace.Wrap(err, "opening listener")
+		}
+		defer func() {
+			if err := l.Close(); err != nil && !utils.IsUseOfClosedNetworkError(err) {
+				s.log.ErrorContext(ctx, "Failed to close listener", "error", err)
+			}
+		}()
+	}
+
+	lpCfg, err := s.buildLocalProxyConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err, "building local proxy config")
+	}
+	lpCfg.Listener = l
+
+	lp, err := alpnproxy.NewLocalProxy(lpCfg)
+	if err != nil {
+		return trace.Wrap(err, "creating local proxy")
+	}
+	defer func() {
+		if err := lp.Close(); err != nil {
+			s.log.ErrorContext(ctx, "Failed to close local proxy", "error", err)
+		}
+	}()
+	// Closed further down.
+
+	// lp.Start will block and continues to block until lp.Close() is called.
+	// Despite taking a context, it will not exit until the first connection is
+	// made after the context is canceled.
+	var errCh = make(chan error, 1)
+	go func() {
+		errCh <- lp.Start(ctx)
+	}()
+	s.log.InfoContext(ctx, "Listening for connections.", "address", l.Addr().String())
+
+	select {
+	case <-ctx.Done():
+		return nil
+	case err := <-errCh:
+		return trace.Wrap(err, "local proxy failed")
+	}
+}
+
+func alpnProtocolForApp(app types.Application) common.Protocol {
+	if app.IsTCP() {
+		return common.ProtocolTCP
+	}
+	return common.ProtocolHTTP
+}
+
+// buildLocalProxyConfig initializes the service, fetching any initial information and setting
+// up the localproxy.
+func (s *ApplicationTunnelService) buildLocalProxyConfig(ctx context.Context) (lpCfg alpnproxy.LocalProxyConfig, err error) {
+	ctx, span := tracer.Start(ctx, "ApplicationTunnelService/buildLocalProxyConfig")
+	defer span.End()
+
+	// Determine the roles to use for the impersonated app access user. We fall
+	// back to all the roles the bot has if none are configured.
+	roles := s.cfg.Roles
+	if len(roles) == 0 {
+		roles, err = fetchDefaultRoles(ctx, s.botClient, s.getBotIdentity())
+		if err != nil {
+			return alpnproxy.LocalProxyConfig{}, trace.Wrap(err, "fetching default roles")
+		}
+		s.log.DebugContext(ctx, "No roles configured, using all roles available.", "roles", roles)
+	}
+
+	proxyPing, err := s.proxyPingCache.ping(ctx)
+	if err != nil {
+		return alpnproxy.LocalProxyConfig{}, trace.Wrap(err, "pinging proxy")
+	}
+	proxyAddr := proxyPing.Proxy.SSH.PublicAddr
+
+	s.log.DebugContext(ctx, "Issuing initial certificate for local proxy.")
+	appCert, app, err := s.issueCert(ctx, roles)
+	if err != nil {
+		return alpnproxy.LocalProxyConfig{}, trace.Wrap(err)
+	}
+	s.log.DebugContext(ctx, "Issued initial certificate for local proxy.")
+
+	middleware := alpnProxyMiddleware{
+		onNewConnection: func(ctx context.Context, lp *alpnproxy.LocalProxy) error {
+			ctx, span := tracer.Start(ctx, "ApplicationTunnelService/OnNewConnection")
```

Comment:
Probably out of scope for a bunch of reasons - and I doubt `tsh proxy` handles this either - but I noticed while testing this with [`oha`](https://github.com/hatoo/oha) that clients that try to keep connections open will eventually start getting 403s and won't recover. 

Swapping the tunnel out transparently sounds ... hard, but I wonder if we could close connections whose certs expire so clients don't try to reuse them?

Otherwise, it does happily refresh certs for new connections, which is a huge improvement. The 403 semantics just seem a bit weird and might not indicate to clients that the error can be fixed with a retry - not that this behavior is anything new.

Reply:
Hmm yeah - I think definitely something we can revisit. If we know it's a HTTP app, then perhaps we actually can just be a little more aware and re-open an upstream if we need it.